File: Cpp/Lekcja11.cpp
#include <iostream>
#include <cmath>

// Lekcja 11 - Funkcje


// typ nazwa (argumenty) { instrukcje; return (zwracana wartosc) }
// typ void oznacza, że funkcja nie zwraca wartości
void funkcja(int argument) {
    std::cout << "\nPodany argument to: " << argument << std::endl;
}

// funkcja nie musi przyjmowac wartości, ani ich zwracać
void helloWorld() {
    std::cout << "\nHello World!" << std::endl;
}

// poniższa funkcja zwraca nam wynik, ale nie wypisuje go sama
int suma(int a, int b) {
    int wynik = a + b;
    return wynik;
}

// funkcja ma argument domyślny (wykładnik)
// oznacza to, że nie musimy go podawac w jej wywołaniu
int potega(int podstawa, int wykladnik = 1) {
    int wynik = 1;
    for (int i = 0; i < wykladnik; i++) {
        wynik *= podstawa;
    }
    return wynik;
}

// main też jest funkcją
int main()
{
    int a = 5, b = 4, c = 3;
    helloWorld();
    funkcja(a);
    funkcja(50);
    int wynikSumy = suma(b, c);
    std::cout << "\nWynik funkcji suma: " << wynikSumy;
    std::cout << "\nWynik funkcji potegowania: " << potega(b);
    std::cout << "\nWynik funkcji potegowania: " << potega(b, c);

    return 0;
}


// Zadanie 1 - napisz funkcję, która przyjmuje zakres (n)
// i wypisuje wszystkie liczby mieszczące się w nim (od 1 do n-1)

// Zadanie 2 - napisz funkcję, która przyjmuje liczbę
// jeżeli ta liczba będzie ujemna to zwraca false
// a jeżeli dodatnia to true

// Zadanie 3 - napisz funkcję, która przyjmuje dwie liczby
// i zwraca większą z nich

File: Python/lekcja_11.py
# Funkcje II

def silnia(liczba):
    wynik = 1
    for i in range(1, liczba+1):
        wynik = i * wynik
    return wynik

# Poniższa funkcja nie różni się od funkcji silnia
# mamy tu jednak dodany type hinting
# który pozwala zniwelowac błędy związane z niepoprawnymi typami
# wystarczy dopisac jaki typ mają argumenty i jaki typ jest zwracany
def silnia_bezpieczna(liczba: int) -> int:
    wynik = 1
    for i in range(1, liczba+1):
        wynik = i * wynik
    return wynik


#print(silnia_bezpieczna(5))
#print(silnia("Slowo"))
#print(silnia_bezpieczna("Slowo"))


# Zadanie 1
'''
Przygotuj funkcję, która jako argument otrzymuje string oraz listę stringów, a
zwraca napis, gdzie pomiędzy elementy z listy dokładana jest zawartość
pierwszego argumentu. Przykład:

foo(‘?’, [‘ala’, ‘ma’, ‘kota’]) -> ‘ala?ma?kota’
'''

def foo(symbol: str, slowa: list[str]) -> str:
    slowo_pomocnicze = slowa[0]
    
    for slowo in slowa[1:]:
        slowo_pomocnicze += symbol + slowo
    
    return slowo_pomocnicze

def foo2(symbol: str, slowa: list[str]) -> str:
    return symbol.join(slowa)

print(foo('?', ['ala', 'ma', 'kota']))
print(foo2('?', ['ala', 'ma', 'kota']))

# Zadanie 2
'''
Napisz funkcję, która otrzyma dwa argumenty pierwszym będzie liczba, którą
chcemy podzielić bez reszty a drugim argumentem będzie dzielnik. 
Sprawdź czy dzielnik jest poprawny (to znaczy nie ma reszty z dzielenia)
'''

def dzielenie(liczba: int, dzielnik: int):
    if not (liczba % dzielnik):
        print("Jest to dzielnik tej liczby")
    else:
        print("Nie jest to dzielnik tej liczby")

# Zadanie 3
# Funkcja do logowania, która pozwala na n prób zalogowań
# jeżeli użytkownik wpisze hasło źle więcej niż n razy
# to nie może wpisywać już więcej

# Zadanie 4 
'''
Napisz funkcję przyjmującą imię, wiek i wzrost
i zwraca string "{imie}, lat {wiek}, {wzrost} m wzrostu" 
'''

info = f"informacje"

# nazwa = lambda argumenty : instrukcje
mnozenie = lambda a, b : a * b
print(mnozenie(5, 6))
print(mnozenie(7, 8))

# Zadanie 5 
# Funkcja do wyszukiwania w liczb pierwszych w danym zakresie
# można użyć Sito Eratostenesa

def prime(zakres: int) -> list[int]:
    liczby_pierwsze: list = []
    
    for cyfra in range(2, zakres+1):
        jest_pierwsza = True
        
        for dzielnik in range (2, cyfra):
            if cyfra % dzielnik == 0:
                jest_pierwsza = False
        
        if jest_pierwsza:
            liczby_pierwsze.append(cyfra)
    
    return liczby_pierwsze

print(prime(30))
            